File: Windows/Domain/ExecutionModels.cs
using System.Text.Json.Serialization;

namespace CytisusTrading.Windows;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum OrderSide
{
    Buy,
    Sell
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum RiskDecisionOutcome
{
    Accepted,
    Rejected
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum BrokerOrderState
{
    Accepted,
    Rejected,
    FullyFilled,
    PartiallyFilled,
    Expired,
    Ambiguous
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ReconciliationStatus
{
    Reconciled,
    Mismatch
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum LiveSubmissionState
{
    Disabled,
    CommandConstructed,
    Rejected,
    Submitted,
    Ambiguous
}

public sealed record TradeIntent(
    int SchemaVersion,
    string IntentId,
    string StrategyId,
    string StrategyVersion,
    string CycleId,
    string SettlementCycle,
    string Symbol,
    decimal RequestedQuantity,
    int Priority,
    bool AllowPartial,
    decimal MinimumEffectiveFill,
    int TimeToLiveSeconds,
    string ReasonCode,
    int ParameterVersion,
    DateTimeOffset CreatedAt)
{
    public OrderSide Side =>
        RequestedQuantity >= 0 ? OrderSide.Buy : OrderSide.Sell;
    public decimal AbsoluteQuantity => Math.Abs(RequestedQuantity);
}

public sealed record RiskDecision(
    int SchemaVersion,
    string DecisionId,
    string IntentId,
    string StrategyId,
    string Symbol,
    RiskDecisionOutcome Outcome,
    string Reason,
    string CorrelationId,
    DateTimeOffset CreatedAt);

public sealed record InternalTransfer(
    int SchemaVersion,
    string TransferId,
    string Symbol,
    string SettlementCycle,
    string BuyerStrategyId,
    string SellerStrategyId,
    decimal Quantity,
    decimal ReferencePrice,
    string ReferencePriceSource,
    string BuyerIntentId,
    string SellerIntentId,
    string CorrelationId,
    DateTimeOffset CreatedAt)
{
    public string Display =>
        $"{Symbol} | {SellerStrategyId} -> {BuyerStrategyId} | {Quantity:0.####} @ {ReferencePrice:0.00}";
}

public sealed record BrokerOrder(
    int SchemaVersion,
    string OrderId,
    string CorrelationId,
    string Symbol,
    OrderSide Side,
    decimal Quantity,
    decimal ReferencePrice,
    StrategyMode Mode,
    BrokerOrderState State,
    DateTimeOffset CreatedAt,
    DateTimeOffset ExpiresAt,
    int SubmissionAttempts,
    string StatusMessage)
{
    public string ModeLabel =>
        Mode == StrategyMode.PaperOnly ? "Local Paper" : "Live";
}

public sealed record BrokerFill(
    int SchemaVersion,
    string FillId,
    string OrderId,
    string CorrelationId,
    string Symbol,
    OrderSide Side,
    decimal Quantity,
    decimal Price,
    decimal SlippageEstimate,
    decimal FeeEstimate,
    DateTimeOffset FilledAt);

public sealed record VirtualAllocation(
    int SchemaVersion,
    string AllocationId,
    string FillId,
    string IntentId,
    string StrategyId,
    string Symbol,
    OrderSide Side,
    decimal Quantity,
    decimal Price,
    decimal Fee,
    string CorrelationId,
    DateTimeOffset CreatedAt);

public sealed record AllocationShortfall(
    int SchemaVersion,
    string ShortfallId,
    string IntentId,
    string StrategyId,
    string Symbol,
    decimal RequestedQuantity,
    decimal AllocatedQuantity,
    decimal UnfilledQuantity,
    string Reason,
    string CorrelationId,
    DateTimeOffset CreatedAt);

public sealed record VirtualLedgerPosition(
    int SchemaVersion,
    string StrategyId,
    string Symbol,
    decimal TargetPosition,
    decimal VirtualQuantity,
    decimal CostBasis,
    decimal RealizedPnl,
    decimal UnrealizedPnl,
    decimal CapitalUsage,
    decimal RiskContribution,
    IReadOnlyList<string> IntentIds,
    IReadOnlyList<string> AllocationIds,
    IReadOnlyList<string> InternalTransferIds,
    DateTimeOffset UpdatedAt)
{
    public string PositionKey => $"{StrategyId}|{Symbol}";
}

public sealed record ReconciliationEvent(
    int SchemaVersion,
    string ReconciliationId,
    string Symbol,
    decimal VirtualQuantity,
    decimal BrokerQuantity,
    decimal Difference,
    ReconciliationStatus Status,
    bool BlocksNewRisk,
    string Diagnostic,
    string CorrelationId,
    DateTimeOffset CreatedAt);

public sealed record CriticalRiskEvent(
    int SchemaVersion,
    string RiskEventId,
    string Symbol,
    string Severity,
    string Message,
    bool Persistent,
    string CorrelationId,
    DateTimeOffset CreatedAt);

public sealed record ExecutionStateSnapshot(
    int SchemaVersion,
    IReadOnlyList<TradeIntent> Intents,
    IReadOnlyList<RiskDecision> RiskDecisions,
    IReadOnlyList<InternalTransfer> InternalTransfers,
    IReadOnlyList<BrokerOrder> BrokerOrders,
    IReadOnlyList<BrokerFill> BrokerFills,
    IReadOnlyList<VirtualAllocation> VirtualAllocations,
    IReadOnlyList<AllocationShortfall> AllocationShortfalls,
    IReadOnlyList<VirtualLedgerPosition> LedgerPositions,
    IReadOnlyList<ReconciliationEvent> Reconciliations,
    IReadOnlyList<CriticalRiskEvent> RiskEvents,
    IReadOnlyList<string> BlockedSymbols)
{
    public static ExecutionStateSnapshot Empty { get; } = new(
        1,
        Array.Empty<TradeIntent>(),
        Array.Empty<RiskDecision>(),
        Array.Empty<InternalTransfer>(),
        Array.Empty<BrokerOrder>(),
        Array.Empty<BrokerFill>(),
        Array.Empty<VirtualAllocation>(),
        Array.Empty<AllocationShortfall>(),
        Array.Empty<VirtualLedgerPosition>(),
        Array.Empty<ReconciliationEvent>(),
        Array.Empty<CriticalRiskEvent>(),
        Array.Empty<string>());
}

public sealed record ExecutionGatewayContext(
    StrategyMode Mode,
    bool GlobalLiveLock,
    LiveAuthorization? Authorization,
    bool DataFresh,
    bool MarketAllowed,
    HealthState StrategyHealth,
    decimal CapitalBudget,
    decimal MaximumPositionExposure,
    bool CliReady,
    bool LiveAdapterEnabled,
    bool FixtureMode,
    int ParameterVersion,
    string Market,
    DateTimeOffset Now,
    decimal DailyLoss = 0,
    decimal Drawdown = 0,
    bool OutsideRegularHours = false,
    bool TransitionActive = false,
    LiveToPaperTransition TransitionPolicy =
        LiveToPaperTransition.StopOpeningRisk);

public sealed record GatewayBatchResult(
    ExecutionStateSnapshot State,
    IReadOnlyList<RiskDecision> Decisions,
    IReadOnlyList<InternalTransfer> Transfers,
    IReadOnlyList<BrokerOrder> Orders,
    IReadOnlyList<BrokerFill> Fills,
    IReadOnlyList<VirtualAllocation> Allocations,
    IReadOnlyList<AllocationShortfall> Shortfalls,
    IReadOnlyList<ReconciliationEvent> Reconciliations);

public sealed record PaperBrokerConfiguration(
    decimal FillRatio,
    decimal SlippageBasisPoints,
    decimal FeePerUnit,
    decimal MinimumFee,
    bool RejectOrders,
    bool ExpireOrders);

public sealed record BrokerSubmissionResult(
    BrokerOrder Order,
    BrokerFill? Fill);

public sealed record LongbridgeExecutionCapability(
    int SchemaVersion,
    string SourceVersion,
    bool SyntheticFixture,
    bool SupportsMachineReadableOutput,
    bool SupportsOrderSubmission,
    IReadOnlyList<string> ArgumentTemplate,
    string Message);

public sealed record LiveAdapterConfiguration(
    bool Enabled,
    bool FixtureMode,
    string ExecutablePath,
    TimeSpan Timeout,
    LongbridgeExecutionCapability Capability);

public sealed record LiveSubmissionResult(
    LiveSubmissionState State,
    BrokerOrder Order,
    IReadOnlyList<string> Arguments,
    bool RetryPermitted,
    string Message);

public sealed record FixtureLedgerSeed(
    string StrategyId,
    decimal VirtualQuantity,
    decimal CostBasis);

public sealed record PaperGatewayFixture(
    int SchemaVersion,
    DateTimeOffset AsOf,
    string Market,
    string Symbol,
    decimal ReferencePrice,
    string ReferencePriceSource,
    decimal StartingBrokerQuantity,
    decimal PaperFillRatio,
    IReadOnlyList<FixtureLedgerSeed> StartingLedger,
    IReadOnlyList<TradeIntent> Intents);

public sealed record ReconciliationFailureFixture(
    int SchemaVersion,
    DateTimeOffset AsOf,
    string Symbol,
    decimal VirtualQuantity,
    decimal BrokerQuantity,
    ReconciliationStatus ExpectedStatus,
    bool ExpectedBlocksNewRisk);

File: Sources/Domain/ExecutionModels.swift
import Foundation

enum OrderSide: String, Codable {
    case buy = "Buy"
    case sell = "Sell"
}

enum RiskDecisionOutcome: String, Codable {
    case accepted = "Accepted"
    case rejected = "Rejected"
}

enum BrokerOrderState: String, Codable {
    case accepted = "Accepted"
    case rejected = "Rejected"
    case fullyFilled = "FullyFilled"
    case partiallyFilled = "PartiallyFilled"
    case expired = "Expired"
    case ambiguous = "Ambiguous"
}

enum ReconciliationStatus: String, Codable {
    case reconciled = "Reconciled"
    case mismatch = "Mismatch"
}

enum LiveSubmissionState: String, Codable {
    case disabled = "Disabled"
    case commandConstructed = "CommandConstructed"
    case rejected = "Rejected"
    case submitted = "Submitted"
    case ambiguous = "Ambiguous"
}

struct TradeIntent: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let intentId: String
    let strategyId: String
    let strategyVersion: String
    let cycleId: String
    let settlementCycle: String
    let symbol: String
    let requestedQuantity: Double
    let priority: Int
    let allowPartial: Bool
    let minimumEffectiveFill: Double
    let timeToLiveSeconds: Int
    let reasonCode: String
    let parameterVersion: Int
    let createdAt: Date

    var id: String { intentId }
    var side: OrderSide { requestedQuantity >= 0 ? .buy : .sell }
    var absoluteQuantity: Double { abs(requestedQuantity) }
}

struct RiskDecision: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let decisionId: String
    let intentId: String
    let strategyId: String
    let symbol: String
    let outcome: RiskDecisionOutcome
    let reason: String
    let correlationId: String
    let createdAt: Date

    var id: String { decisionId }
}

struct InternalTransfer: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let transferId: String
    let symbol: String
    let settlementCycle: String
    let buyerStrategyId: String
    let sellerStrategyId: String
    let quantity: Double
    let referencePrice: Double
    let referencePriceSource: String
    let buyerIntentId: String
    let sellerIntentId: String
    let correlationId: String
    let createdAt: Date

    var id: String { transferId }
    var display: String {
        "\(symbol) | \(sellerStrategyId) -> \(buyerStrategyId) | " +
            String(format: "%.4f @ %.2f", quantity, referencePrice)
    }
}

struct BrokerOrder: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let orderId: String
    let correlationId: String
    let symbol: String
    let side: OrderSide
    let quantity: Double
    let referencePrice: Double
    let mode: StrategyMode
    let state: BrokerOrderState
    let createdAt: Date
    let expiresAt: Date
    let submissionAttempts: Int
    let statusMessage: String

    var id: String { orderId }
    var modeLabel: String {
        mode == .paperOnly ? "Local Paper" : "Live"
    }
}

struct BrokerFill: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let fillId: String
    let orderId: String
    let correlationId: String
    let symbol: String
    let side: OrderSide
    let quantity: Double
    let price: Double
    let slippageEstimate: Double
    let feeEstimate: Double
    let filledAt: Date

    var id: String { fillId }
}

struct VirtualAllocation: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let allocationId: String
    let fillId: String
    let intentId: String
    let strategyId: String
    let symbol: String
    let side: OrderSide
    let quantity: Double
    let price: Double
    let fee: Double
    let correlationId: String
    let createdAt: Date

    var id: String { allocationId }
}

struct AllocationShortfall: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let shortfallId: String
    let intentId: String
    let strategyId: String
    let symbol: String
    let requestedQuantity: Double
    let allocatedQuantity: Double
    let unfilledQuantity: Double
    let reason: String
    let correlationId: String
    let createdAt: Date

    var id: String { shortfallId }
}

struct VirtualLedgerPosition: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let strategyId: String
    let symbol: String
    let targetPosition: Double
    let virtualQuantity: Double
    let costBasis: Double
    let realizedPnl: Double
    let unrealizedPnl: Double
    let capitalUsage: Double
    let riskContribution: Double
    let intentIds: [String]
    let allocationIds: [String]
    let internalTransferIds: [String]
    let updatedAt: Date

    var id: String { "\(strategyId)|\(symbol)" }
}

struct ReconciliationEvent: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let reconciliationId: String
    let symbol: String
    let virtualQuantity: Double
    let brokerQuantity: Double
    let difference: Double
    let status: ReconciliationStatus
    let blocksNewRisk: Bool
    let diagnostic: String
    let correlationId: String
    let createdAt: Date

    var id: String { reconciliationId }
}

struct CriticalRiskEvent: Codable, Identifiable, Equatable {
    let schemaVersion: Int
    let riskEventId: String
    let symbol: String
    let severity: String
    let message: String
    let persistent: Bool
    let correlationId: String
    let createdAt: Date

    var id: String { riskEventId }
}

struct ExecutionStateSnapshot: Codable, Equatable {
    let schemaVersion: Int
    var intents: [TradeIntent]
    var riskDecisions: [RiskDecision]
    var internalTransfers: [InternalTransfer]
    var brokerOrders: [BrokerOrder]
    var brokerFills: [BrokerFill]
    var virtualAllocations: [VirtualAllocation]
    var allocationShortfalls: [AllocationShortfall]
    var ledgerPositions: [VirtualLedgerPosition]
    var reconciliations: [ReconciliationEvent]
    var riskEvents: [CriticalRiskEvent]
    var blockedSymbols: Set<String>

    static let empty = ExecutionStateSnapshot(
        schemaVersion: 1,
        intents: [],
        riskDecisions: [],
        internalTransfers: [],
        brokerOrders: [],
        brokerFills: [],
        virtualAllocations: [],
        allocationShortfalls: [],
        ledgerPositions: [],
        reconciliations: [],
        riskEvents: [],
        blockedSymbols: []
    )
}

struct ExecutionGatewayContext {
    let mode: StrategyMode
    let globalLiveLock: Bool
    let authorization: LiveAuthorization?
    let dataFresh: Bool
    let marketAllowed: Bool
    let strategyHealth: HealthState
    let capitalBudget: Double
    let maximumPositionExposure: Double
    let cliReady: Bool
    let liveAdapterEnabled: Bool
    let fixtureMode: Bool
    let parameterVersion: Int
    let market: String
    let now: Date
    let dailyLoss: Double
    let drawdown: Double
    let outsideRegularHours: Bool
    let transitionActive: Bool
    let transitionPolicy: LiveToPaperTransition
}

struct GatewayBatchResult {
    let state: ExecutionStateSnapshot
    let decisions: [RiskDecision]
    let transfers: [InternalTransfer]
    let orders: [BrokerOrder]
    let fills: [BrokerFill]
    let allocations: [VirtualAllocation]
    let shortfalls: [AllocationShortfall]
    let reconciliations: [ReconciliationEvent]
}

struct PaperBrokerConfiguration {
    let fillRatio: Double
    let slippageBasisPoints: Double
    let feePerUnit: Double
    let minimumFee: Double
    let rejectOrders: Bool
    let expireOrders: Bool
}

struct BrokerSubmissionResult {
    let order: BrokerOrder
    let fill: BrokerFill?
}

struct LongbridgeExecutionCapability: Codable, Equatable {
    let schemaVersion: Int
    let sourceVersion: String
    let syntheticFixture: Bool
    let supportsMachineReadableOutput: Bool
    let supportsOrderSubmission: Bool
    let argumentTemplate: [String]
    let message: String
}

struct LiveAdapterConfiguration {
    let enabled: Bool
    let fixtureMode: Bool
    let executableURL: URL?
    let timeout: TimeInterval
    let capability: LongbridgeExecutionCapability
}

struct LiveSubmissionResult {
    let state: LiveSubmissionState
    let order: BrokerOrder
    let arguments: [String]
    let retryPermitted: Bool
    let message: String
}

struct FixtureLedgerSeed: Codable, Equatable {
    let strategyId: String
    let virtualQuantity: Double
    let costBasis: Double
}

struct PaperGatewayFixture: Codable, Equatable {
    let schemaVersion: Int
    let asOf: Date
    let market: String
    let symbol: String
    let referencePrice: Double
    let referencePriceSource: String
    let startingBrokerQuantity: Double
    let paperFillRatio: Double
    let startingLedger: [FixtureLedgerSeed]
    let intents: [TradeIntent]
}

struct ReconciliationFailureFixture: Codable, Equatable {
    let schemaVersion: Int
    let asOf: Date
    let symbol: String
    let virtualQuantity: Double
    let brokerQuantity: Double
    let expectedStatus: ReconciliationStatus
    let expectedBlocksNewRisk: Bool
}
